### ISSUE TITLE
Rancher Ruleset not synchronising beyond ~300 containers / Migrating from @rawmind0 traefik rancher template

### ISSUE THREAD

**Issue by PicoCreator** [traefik/traefik#2125]
### Do you want to request a feature or report a bug?

Bug

### What did you do?

Docker System : Rancher v1.6.5
Server : 3 x AWS r4.8xlarge
Traefik Setup : 1 for each server

Number of node.js containers : ~300
Number of domains per node.js container : 2
- 1 which is the default generated by service name
- 1 by a unique URL, generated for the node.js web instances via 'traefik.alias.fqdn'

We have a webservice which spins up node.js containers for each customer. 
We noticed that beyond approximately ~300 containers, the ~"301" container
does not show up on the admin panel, nor gets accessible.

We have attempted to restart all traefik instances, loadbalancers, and even node.js server.
We confirmed healthcheck is valid on rancher
We replicated the issue, by removing instances, and adding them back.
Instance 1-250+ is consistently working

Not sure if this issue is limited by rancher side, or traefik side

Note: 300 is not the exact number, but a close approximate.

### What did you expect to see?

All node.js web instances to be on the admin panel, and that all node.js web instances 'traefik.alias.fqdn' is properly routed

### What did you see instead?

Missin container entries, 404 for fqdn URL

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.3.6
Codename:     raclette
Go version:   go1.8.3
Built:        2017-08-22_09:37:23AM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Using the community catalogue for rancher : https://github.com/rancher/community-catalog/tree/master/templates/traefik

Note that features such as SSL cert setup, is not used. Traefik is behind an SSL loadbalancer with a wildcard certificate.

**Comment by ldez** [comment]
Could so try with v1.4.0-rc2?

**Comment by PicoCreator** [comment]
@ldez : Currently 1.4.0 isn't an option in production environment.

Im going to try replicate this in our development environment, with hello-world web containers. But that will take some time.

**Comment by SantoDE** [comment]
Hey @PicoCreator ,

the version you're using is not "our" integration with rancher. It's a custom integration made by @rawmind0. 

You could try to deploy "our" API (in the stable version, it's by API access). Please have a look at the [docs](https://docs.traefik.io/configuration/backends/rancher/).

**Comment by PicoCreator** [comment]
@SantoDE, sure will do up a new setup using : https://github.com/containous/traefik-library-image/blob/57f83af9d2719e5f25e6c769fd2640f973b519dc/alpine/Dockerfile

**Comment by PicoCreator** [comment]
## TLDR: its working, for API mode, for metadata mode, sort of ...

Currently in metadata mode backend, services are registered using "container-name" instead of "stack-name"-"container-name" by default. 

This may cause namespace collisions for some backend, which it did for my use case 
(will file this as seperate issue)

## Testing traefik configuration

Because the documentation isn't fully updated for rancher deployment. Im gonna do a quick run through of what I done, for easy reference.

Deployed latest official image : `traefik:v1.4.0-rc2-alpine`

With the following command for API mode (formatted to be easier to read, confidential information been replaced)
```
--web --rancher 
	--rancher.domain=rancher.localhost 
	--rancher.enableservicehealthfilter=false 
	--rancher.exposedbydefault=false 
	--rancher.api.endpoint=https://rancher.your-infra.something
	--rancher.api.accesskey=SECRET_API_ACCESSKEY 
	--rancher.api.secretkey=SECRET_API_SECRETKEY 
	--rancher.refreshseconds=10
```

Or the following command for meta-data mode (formatted to be easier to read)

```
--web --rancher 
	--rancher.domain=rancher.localhost 
	--rancher.enableservicehealthfilter=false 
	--rancher.exposedbydefault=false 
	--rancher.metadata
```

For meta-data mode, you will need to ensure the container is set with the following labels

```
io.rancher.container.agent.role=environment
io.rancher.container.create_agent=true
```

## Migrating from rawmind0 template

As the `traefik.alias.fqdn` in rawmind0 template does not officially exists the following docker container labels in rancher need to be converted from

`traefik.alias.fqdn: hello-traefik-test-1.your-test-domain.internal`

to

`traefik.frontend.rule: "Host:hello-traefik-test-1.your-test-domain.internal"`

## Setting up 600 hello worlds

![screen shot 2017-09-18 at 12 31 06 pm](https://user-images.githubusercontent.com/17175484/30529182-50f18a90-9c6d-11e7-9e07-c876a7fea0ed.png)

Warning : Spinning up 600 servers at a time, can crash your rancher management server if your running it a t2.micro or something like that. And will most definitely cause your rancher UI to lag, in infrastructure page. Or stack page when its open. Also while, each hello-world takes only 2MB of ram, this would mean over a GB of ram. So make sure your testing infra has the capacity.

https://gist.github.com/PicoCreator/baf3a5ea612d003ef36a86b3fd3b5d34

The following docker-compose was generated, for setting up 600 hello world servers with traefik configurations. Note that this is setup to be compatible with both @rawmind0 setup and official configuration.

To actually use this, search and replace the "your-test-domain.internal" with the actual domain

## Testing 600 hello worlds

Manually testing 600 hello worlds would have been a pain, so as a web testing automation company, it was time to do some dog-feeding.

Disclosure : I am a co-founder of uilicious.com

![screen shot 2017-09-17 at 2 02 18 am](https://user-images.githubusercontent.com/17175484/30529393-7cac06cc-9c6f-11e7-87ee-c626803aba20.png)

Naturally as you may have suspected, the rawmind0 integration was breaking on generating the 600+ rulesets.

---------------

**PS:** If it helps, I can always do up a user guide for rancher

**Comment by SantoDE** [comment]
Hi :) Thanks for your feedback and you're very solid testing / summary.

If you feel you can and want to enhance our documentation, we will happily accept a PR. Thanks  :)

**Comment by PicoCreator** [comment]
@SantoDE @ldez , Sorry to reopen the issue. Sad to say after more experimentation and testing I realised the problem is still valid. 

Instead of spinning up 600 servers in a single "Stack", i setup 300 stacks with 1 container each. With the number 10001 - 10300, so that they will appear properly in sequence on traefik.

![screen shot 2017-09-19 at 3 03 07 pm](https://user-images.githubusercontent.com/17175484/30579550-c06ba976-9d4b-11e7-82d6-16b1b9802a92.png)

Unfortunately for the API mode, it only registered up to stack "95". With stack "96" onwards going missing

![screen shot 2017-09-19 at 2 59 02 pm](https://user-images.githubusercontent.com/17175484/30579410-2dcbd168-9d4b-11e7-94f4-e6afb7bfdcf1.png)

Good news however, for meta-data mode, it registers all 300 frontends, and with https://github.com/containous/traefik/pull/2107 tested in rc3 (thanks once again)

![screen shot 2017-09-19 at 2 56 59 pm](https://user-images.githubusercontent.com/17175484/30579513-986095ea-9d4b-11e7-9d05-a861cc8fc400.png)

Note : Traefik configuration is the same as previous test, but in a brand new environment. And updated to  `traefik:v1.4.0-rc3-alpine`

**Comment by SantoDE** [comment]
Hey @PicoCreator,

Thanks for reporting. I will have a look :)

**Comment by PicoCreator** [comment]
@SantoDE 👍  let me know if you want me to test any fixes

**Comment by dbsanfte** [comment]
Rancher's API paginates queries beyond 100 results, I suspect this is likely what you're seeing. 

Query string on the metadata service just needs to be ?limit=-1 or similar to get all the results.

**Comment by ldez** [comment]
@dbsanfte I'm not sure because:
- https://github.com/rancher/api-spec/blob/master/specification.md#pagination

https://github.com/containous/traefik/blob/c0563f1a395d53edbbf1030b2387435077eeb871/provider/rancher/api.go#L33-L35

https://github.com/containous/traefik/blob/c0563f1a395d53edbbf1030b2387435077eeb871/provider/rancher/api.go#L141

https://github.com/containous/traefik/blob/c0563f1a395d53edbbf1030b2387435077eeb871/provider/rancher/api.go#L158

https://github.com/containous/traefik/blob/c0563f1a395d53edbbf1030b2387435077eeb871/provider/rancher/api.go#L175

**Comment by chris-scentregroup** [comment]
The without pagination usage in last comment is correct. For go-rancher that is what should be currently used. Not sure where this might need to be fixed but, default pagination is 100 iirc.

**Comment by tfny** [comment]
Hey y'all, 
Because we don't support version one anymore, I am going to go ahead and close this issue.  Thanks!
